app/models/template.rb

```diff
@@ -73,8 +73,99 @@ def ignore_default
     self
   end
 
+  # Set attributes based on template +text and metadata found in this +text
+  # the metadata parsing can be adjusted using +options
+  def import_without_save(text, options = {})
+    self.template = text
+    @importing_metadata = self.class.parse_metadata(text)
+    Foreman::Logging.logger('app').debug "setting attributes for #{self.name} with id: #{self.id || 'N/A'}"
+    self.snippet = !!@importing_metadata[:snippet]
+
+    import_locations(options)
+    import_organizations(options)
+    import_custom_data(options)
+
+    self
+  end
+
+  # Set template attributes
+  #
+  # based on +name it either finds existing template or builds a new one
+  # then it applies changes to it and return this object, note no changes were saved at this point
+  def self.import_without_save(name, text, options = {})
+    template = self.unscoped.find_or_initialize_by(:name => name)
```

Comment:
This smells like cross-org template injection. I haven't tried yet, but if there is a scenario with existing template and an attacker know the name, template text field from different organization can be replaced possibly. We only check taxonomy for fields sent via metadata, but if they are not present, the check is skipped. Let me show down below in `import_organizations`.

Reply:
This is in fact covered by permission check on creation and taxonomy validation. We can't have two objects with the same name, taxonomix does not handle it and it would fail on mísleading validation error. So if such template exists, we use it as the object, but when you try to save it, it would fail with message like "you're not authorized for such update" if that would result in escalation. If it's not clear from the comment, I suggest you test it. Also keep in mind that non-adminuser can't create template without choosing org and loc.